PSPs/drafts/psp-30.md

```diff
@@ -0,0 +1,882 @@
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "name": "from",
+      "type": {
+        "displayName": [
+          "AccountId"
+        ],
+        "type": "AccountId"
+      }
+    },
+    {
+      "name": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    },
+    {
+      "name": "data",
+      "type": {
+        "displayName": [
+          "[u8]"
+        ],
+        "type": "[u8]"
+      }
+    }
+  ],
+  "docs": [
+    " Ensures that the smart contract allows reception of PSP30 token(s).",
+    " Returns `Ok(())` if the contract allows the reception of the token(s) and Error `TransferRejected(String))` otherwise.",
+    "",
+    " This method will get called on every transfer to check whether the recipient in `transfer`",
+    " or `transfer_from` is a contract, and if it is, does it accept tokens.",
+    " This is done to prevent contracts from locking tokens forever.",
+    "",
+    " Returns `PSP30ReceiverError` if the contract does not accept the tokens."
+  ],
+  "mutates": true,
+  "name": [
+    "PSP30Receiver",
+    "before_received"
+  ],
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Result"
+    ],
+    "type": 2
+  },
+  "selector": "" // TO UPDATE WHEN PSP NUMBER
+}
+```
+
+### Extension
+
+#### PSP30Metadata
+
+`PSP30Metadata` is an optional interface for metadata for this Non-Fungible Token standard.
+
+##### **name**() ➔ Option<String>
+Selector: `` - first 4 bytes of `blake2b_256("")` // TO UPDATE WHEN PSP NUMBER
+```json
+{
+  "args": [],
+  "docs": [
+    "Returns the token name."
+  ],
+  "mutates": false,
+  "name": [
+    "PSP30Metadata",
+    "name"
+  ],
+  "returnType": {
+    "displayName": [
+      "Option"
+    ],
+    "type": "Option<string>"
+  },
+  "selector": "" // TO UPDATE WHEN PSP NUMBER
+}
+```
+
+##### **symbol**() ➔ Option<String>
+Selector: `` - first 4 bytes of `blake2b_256("")` // TO UPDATE WHEN PSP NUMBER
+```json
+{
+  "args": [],
+  "docs": [
+    "Returns the token symbol."
+  ],
+  "mutates": false,
+  "name": [
+    "PSP22Metadata",
+    "symbol"
+  ],
+  "returnType": {
+    "displayName": [
+      "Option"
+    ],
+    "type": "Option<string>"
+  },
+  "selector": "" // TO UPDATE WHEN PSP NUMBER
+}
+```
+
+##### **uri**(id: Id) ➔ Option<String>
+Selector: `` - first 4 bytes of `blake2b_256("")` // TO UPDATE WHEN PSP NUMBER
+```json
+{
+  "args": [
+    {
+      "name": "id",
+      "type": {
+        "displayName": [
+           "Id"
+        ],
+        "type": "Id"
+      }
+    }
+  ],
+  "docs": [
+    " Returns the Uniform Resource Identifier (URI) for `id` token."
+  ],
+  "mutates": false,
+  "name": [
+    "PSP30Metadata",
+    "uri"
+  ],
+  "payable": false,
+  "returnType": {
+    "displayName": [
+      "Option"
+    ],
+    "type": "Option<string>"
+  },
+  "selector": "" // TO UPDATE WHEN PSP NUMBER
+}
+```
+
+### Events
```

Comment:
Also, I think we should add a new event for `URI`.

```
    /**
        @dev MUST emit when the URI is updated for a token ID.
        URIs are defined in RFC 3986.
        The URI MUST point to a JSON file that conforms to the "ERC-721 Metadata URI JSON Schema".
    */
    event URI(string _value, uint256 indexed _id);
```
The contract should emit that event each time when URI for some `Id` is changed. The id should be `Optional` here. If the event is emitted with `None`, it means that URI has changed for all ids.